client/src/event.rs

```diff
@@ -0,0 +1,176 @@
+        } else {
+            let records = Vec::<event_v2::Record>::decode(&mut data)?;
+            Ok(records.into_iter().map(EventRecord::V2).collect())
+        }
+    }
+
+    pub fn event(&self) -> Event {
+        match self {
+            EventRecord::V1(record) => Event::V1(record.event.clone()),
+            EventRecord::V2(record) => Event::V2(record.event.clone()),
+        }
+    }
+
+    /// Return the index of the transaction in the block that dispatched the event.
+    ///
+    /// Returns `None` if the event was not dispatched as part of a transaction.
+    pub fn transaction_index(&self) -> Option<u32> {
+        match self {
+            EventRecord::V1(record) => event_v1::transaction_index(record),
+            EventRecord::V2(record) => event_v2::transaction_index(record),
+        }
+    }
+}
+
+/// Compatibility adapter for runtime [EventV2].
+pub enum Event {
+    V1(EventV1),
+    V2(EventV2),
+}
+
+impl Event {
+    /// Extracts the extrinsic result from the event.
+    ///
+    /// If the event is either `ExtrinsicSuccess` or `ExtrinsicFailed` it returns `Ok` or the
+    /// `Err`, respectively. If the event is neither of those it returns `None`.
+    pub fn extrinsic_result(&self) -> Option<Result<(), DispatchError>> {
+        match self {
+            Event::V1(event) => match event {
+                EventV1::system(event) => match event {
+                    event_v1::System::ExtrinsicSuccess(_) => Some(Ok(())),
+                    event_v1::System::ExtrinsicFailed(dispatch_error, _) => {
+                        // The SCALE representation of `DisptachError` has not changed between the
+                        // two runtime versions.
+                        let dispatch_error = sp_runtime::DispatchError::decode(
+                            &mut dispatch_error.encode().as_ref(),
+                        )
+                        .expect("");
+                        Some(Err(dispatch_error))
+                    }
+                    _ => None,
+                },
+                _ => None,
+            },
+            Event::V2(event) => match event {
+                EventV2::system(event) => match event {
+                    event_v2::System::ExtrinsicSuccess(_) => Some(Ok(())),
+                    event_v2::System::ExtrinsicFailed(dispatch_error, _) => {
+                        Some(Err(*dispatch_error))
+                    }
+                    _ => None,
+                },
+                _ => None,
+            },
+        }
+    }
+
+    /// Returns the inner registry event if the runtime event is a registry event
+    pub fn registry(&self) -> Option<event_v2::Registry> {
+        match self {
+            Event::V1(EventV1::registry(event)) => Some(
+                // The SCALE representation of a registry event has not changed between the two
+                // runtime versions.
+                event_v2::Registry::decode(&mut event.encode().as_ref())
```

Comment:
```suggestion
                event_v2::Registry::decode_all(&event.encode())
```
Which will require
```rust
use parity_scale_codec::DecodeAll;
```
This makes SCALE validation as good as possible